ui/job-view/details/tabs/TabsPanel.jsx

```diff
@@ -0,0 +1,240 @@
+
+import { thEvents } from '../../../js/constants';
+import { getAllUrlParams } from '../../../helpers/location';
+import { getStatus } from '../../../helpers/job';
+
+import JobDetailsTab from './JobDetailsTab';
+import FailureSummaryTab from './failureSummary/FailureSummaryTab';
+import PerformanceTab from './PerformanceTab';
+import AutoclassifyTab from './autoclassify/AutoclassifyTab';
+import AnnotationsTab from './AnnotationsTab';
+import SimilarJobsTab from './SimilarJobsTab';
+
+export default class TabsPanel extends React.Component {
+  static getDerivedStateFromProps(props, state) {
+    const { perfJobDetail, selectedJob } = props;
+    const { showAutoclassifyTab } = state;
+
+    // This fires every time the props change.  But we only want to figure out the new default
+    // tab when we get a new job.  However, the job could change, then later, the perf details fetch
+    // returns.  So we need to check for a change in the size of the perfJobDetail too.
+    if (state.jobId !== selectedJob.id || state.perfJobDetailSize !== perfJobDetail.length) {
+      const tabIndex = TabsPanel.getDefaultTabIndex(
+        getStatus(selectedJob),
+        !!perfJobDetail.length, showAutoclassifyTab
+      );
+
+      return {
+        tabIndex,
+        jobId: selectedJob.id,
+        perfJobDetailSize: perfJobDetail.length,
+      };
+    }
+    return {};
+  }
+
+  static getTabNames(showPerf, showAutoclassify) {
+    return [
+      'details', 'failure', 'autoclassify', 'annotations', 'similar', 'perf'
+    ].filter(name => (
+      !((name === 'autoclassify' && !showAutoclassify) || (name === 'perf' && !showPerf))
+    ));
+  }
+
+  static getDefaultTabIndex(status, showPerf, showAutoclassify) {
+    let idx = 0;
+    const tabNames = TabsPanel.getTabNames(showPerf, showAutoclassify);
+    const tabIndexes = tabNames.reduce((acc, name) => ({ ...acc, [name]: idx++ }), {});
+
+    let tabIndex = showPerf ? tabIndexes.perf : tabIndexes.details;
+    if (['busted', 'testfailed', 'exception'].includes(status)) {
+      tabIndex = showAutoclassify ? tabIndexes.autoclassify : tabIndexes.failure;
+    }
+    return tabIndex;
+  }
+
+  constructor(props) {
+    super(props);
+
+    const { $injector } = this.props;
+    this.$rootScope = $injector.get('$rootScope');
+
+    this.state = {
+      showAutoclassifyTab: getAllUrlParams().has('autoclassify'),
+      tabIndex: 0,
+      perfJobDetailSize: 0,
+      jobId: null,
+    };
+  }
+
+  componentDidMount() {
+    this.selectNextTabUnlisten = this.$rootScope.$on(thEvents.selectNextTab, () => {
+      const { tabIndex, showAutoclassifyTab } = this.state;
+      const { perfJobDetail } = this.props;
+      const nextIndex = tabIndex + 1;
+      const tabCount = TabsPanel.getTabNames(!!perfJobDetail.length, showAutoclassifyTab).length;
+      this.setState({ tabIndex: nextIndex < tabCount ? nextIndex : 0 });
+    });
+
+    this.setTabIndex = this.setTabIndex.bind(this);
```

Comment:
I'm curious why this method is bound here rather than in the constructor.

Reply:
This appears (as far as I can tell) to be a quirk of mixing angular with react.  In some cases, the function gets bound to a state of the app where you get an error on the console like this:

    Warning: Can't call setState on a component that is not yet mounted. This is a no-op, but it might indicate a bug in your application. Instead, assign to `this.state` directly or define a `state = {};` class property with the desired state in the TabsPanel component. warning.js:33 

So putting it in ``componentDidMount`` hits it at a point where it obviously WAS mounted already.  I'm hoping to move these all into the constructors when we no longer need ``react2angular``.  

I actually referenced this in one of my TODOs here a while back:  https://github.com/mozilla/treeherder/blob/44e25f7ac148a4a7a4fb1a41350d908534ed6ad3/ui/job-view/details/tabs/autoclassify/AutoclassifyTab.jsx#L87

Reply:
Ah, I see. Thanks for explaining :)